### DISCUSSION TITLE
Set currentUser dependency value for child reducer

### DISCUSSION THREAD

**Discussion by shredlocker** [pointfreeco/swift-composable-architecture#2825] [Q&A] [answered]
Dearest community.

I'm trying to integrate a `currentUser` into my application using a dependency.

With my basic understanding of Dependencies my first idea was to have a `currentUser` dependency key
```swift
enum UserState {
    case unknown
    case guest
    case authenticated(Login)
}

extension UserState: DependencyKey {
    static let liveValue: UserState = .unknown
}

extension DependencyValues {
    var currentUser: UserState {
        get { self[UserState.self] }
        set { self[UserState.self] = newValue }
    }
}
```
and then use this dependency within my api client dependency to set authorization header value.
```swift
mutating func prepareHeaders(for endpoint: Endpoint) throws {
        setValue("application/json", forHTTPHeaderField: "Content-Type")
        setValue("application/json", forHTTPHeaderField: "Accept")

        if endpoint.requiresAuthentication {
            @Dependency(\.currentUser) var currentUser
            
            guard case .authenticated(let login) = currentUser else {
                throw URLSessionError.authentication
            }

            setValue("\(login.tokenType) \(login.accessToken)", forHTTPHeaderField: "Authorization")
        }
    }
```

In the `Root` feature reducer I'm using the `Destination` pattern to show either a login view or the home view depending on whether the user is authenticated or not.
```swift
extension Root {
    @Reducer
    enum Destination {
        case authentication(Authentication)
        case home(Home)
    }
}

// Root action
case .loginResult(.success(let login)):
    state.destination = .home(Home.State())
    return .none

// Root View
struct RootView: View {
    @Bindable var store: StoreOf<Root>

    var body: some View {
        ZStack {
            Color.systemBackground
                .ignoresSafeArea()
            
            if let store = store.scope(
                state: \.destination?.authentication,
                action: \.destination.authentication
            ) {
                AuthenticationView(store: store)
            } else if let store = store.scope(
                state: \.destination?.home,
                action: \.destination.home
            ) {
                HomeView(store: store)
            }
        }
        .onAppear {
            store.send(.onAppear)
        }
    }
}
```
Is it possible to override the dependency values for the child `Home` reducer somehow after user has successfully logged in using the destination pattern? Or do I have to use a different pattern for my `currentUser` dependency?
```swift
// Root action
case .loginResult(.success(let login)):
    // Override current user dependency for Home reducer
    ... withDependencies { values in
        values.currentUser = .authenticated(login)
    }
    return .none
```
I'd love to hear suggestions of other patterns of integrating a `currentUser` into a TCA application

**Comment by fabstu** [comment]
`ReducerReader` may be similar to what you want. https://github.com/pointfreeco/swift-composable-architecture/pull/1969

**Comment by mbrandonw** [comment, marked as answer]
Hi @shredlocker, your dependency can hold onto mutable data and you can mutate it though a dependency endpoint. This is how we handle things in the `APIClient` in [isowords](http://github.com/pointfreeco/isowords): 

https://github.com/pointfreeco/isowords/blob/595afc1a425e490f078a7b6e1d4ce6e6f28a56b6/Sources/ApiClient/Client.swift#L7-L15

The `authenticate` endpoint [mutates](https://github.com/pointfreeco/isowords/blob/595afc1a425e490f078a7b6e1d4ce6e6f28a56b6/Sources/ApiClientLive/LiveKey.swift#L83) some internal data to represent the current user, and `logout` [clears out](https://github.com/pointfreeco/isowords/blob/595afc1a425e490f078a7b6e1d4ce6e6f28a56b6/Sources/ApiClientLive/LiveKey.swift#L88) that data.

I think that is typically a better approach than `ReducerReader`.

**Reply by shredlocker** [reply]
This is an interesting approach! I'm still trying to adjust my brain to use these closure values and I didn't consider using them to store some value like that. It also makes sense to make the login info a part of the api client itself and not a separate dependency, like a `userStore`. 

I'll look into using this approach. Thanks for quick answer @mbrandonw

**Comment by omaralbeik** [comment]
@mbrandonw Thank you for your answer

Handling mutable data through a dependency endpoint introduces significant complexity. Let me provide a more detailed example:

I have a client that uses a `Source` type to perform various operations, such as fetching data from an API:

```swift
public struct MediaClient: Sendable {
  public var getSourceInfo: () async throws -> SourceInfo
  public var getChannelCategories: () async throws -> [Category]
  public var getMoviesCategories: () async throws -> [Category]
  public var getShowsCategories: () async throws -> [Category]
  public var getChannels: (Category.ID) async throws -> [Channel]
  public var getMovies: (Category.ID) async throws -> [Movie]
  public var getShows: (Category.ID) async throws -> [Show]
  ...
}
```

The `Source` is required by all these functions. I have two options: either pass it as an argument to every function or manage it as a mutable property within the client:

```swift
  public var source: Source?
  public var loadSource: (Source) -> Void
```

Both approaches present challenges:

1. Exposing the source property becomes problematic since multiple static instances of the client may exist across different modules (e.g., `MediaClient.live`, `MediaClient.demo`, `MediaClient.stub`).
2.  Adding checks in every function to ensure source is not nil introduces unnecessary complexity.

Additionally, some initialization tasks, such as setting up a database, depend on properties when the client is first created.

I experimented with `ReducerReader`, but it wasn't ideal either, as it created a new client instance whenever any state changed.

Given these considerations, what do you think would be the best approach to solve this issue?

Thanks,
Omar
